index.js

```diff
@@ -70,20 +71,102 @@ function format(text, opts, addAlignmentSize) {
   return str;
 }
 
-function formatRange(text, opts) {
+function findNodeByOffset(node, offset, opts, text) {
+  const start = util.locStart(node);
+  const end = util.locEnd(node);
+  if (start <= offset && offset <= end) {
+    for (const childNode of comments.getSortedChildNodes(node)) {
+      const childResult = findNodeByOffset(childNode, offset, opts, text);
+      if (childResult) {
+        return childResult;
+      }
+    }
+
+    try {
+      if (isSourceElement(node)) {
+        return node;
+      }
+    } catch (err) {
+      return;
+    }
+  }
+}
+
+// See https://www.ecma-international.org/ecma-262/5.1/#sec-A.5
+function isSourceElement(node) {
+  if (node == null) {
+    return false;
+  }
+  switch (node.type) {
+    case "FunctionDeclaration":
+      return true;
+    // Statements
+    case "BlockStatement":
+    case "BreakStatement":
+    case "ContinueStatement":
+    case "DebuggerStatement":
+    case "DoWhileStatement":
+    case "EmptyStatement":
+    case "ExpressionStatement":
+    case "ForInStatement":
+    case "ForStatement":
+    case "IfStatement":
+    case "LabeledStatement":
+    case "ReturnStatement":
+    case "SwitchStatement":
+    case "ThrowStatement":
+    case "TryStatement":
+    case "VariableDeclaration":
+    case "WhileStatement":
+    case "WithStatement":
+      return true;
+  }
+  return false;
+}
+
+function calculateRange(text, opts, ast) {
+  const startNode = findNodeByOffset(ast, opts.rangeStart, opts, text);
+  const endNode = findNodeByOffset(ast, opts.rangeEnd, opts, text);
+  const rangeStart = Math.min(util.locStart(startNode), util.locStart(endNode));
+  const rangeEnd = Math.max(util.locEnd(startNode), util.locEnd(endNode));
+
+  const rangeString = text.slice(rangeStart, rangeEnd);
+
+  // Try to extend the range backwards to the beginning of the line.
+  // This is so we can detect indentation correctly and restore it.
   // Use `Math.min` since `lastIndexOf` returns 0 when `rangeStart` is 0
-  const rangeStart = Math.min(
-    opts.rangeStart,
-    text.lastIndexOf("\n", opts.rangeStart) + 1
+  const rangeStart2 = Math.min(
+    rangeStart,
+    text.lastIndexOf("\n", rangeStart) + 1
   );
-  // Use `text.length - 1` as the maximum since `indexOf` returns -1 if `fromIndex >= text.length`
-  const fromIndex = Math.min(opts.rangeEnd, text.length - 1);
-  const nextNewLineIndex = text.indexOf("\n", fromIndex);
-  const rangeEnd = (nextNewLineIndex < 0 ? fromIndex : nextNewLineIndex) + 1; // Add one to make rangeEnd exclusive
-
-  if (0 < rangeStart || rangeEnd < text.length) {
-    const rangeString = text.substring(rangeStart, rangeEnd);
-    const alignmentSize = getAlignmentSize(
+  const indentString = text.slice(rangeStart2, rangeStart);
+  try {
+    if (indentString.trim() == "") {
+      return {
+        rangeStart: rangeStart2,
+        rangeEnd: rangeEnd,
+        rangeString: text.slice(rangeStart2, rangeEnd)
+      };
+    }
+  } catch (err) {
```

Comment:
Thanks, fixed in https://github.com/prettier/prettier/pull/1659/commits/d9abe4b3ff2411b9a5afdaa52399ab8b6700a10e